cloudify_agent/api/plugins/installer.py

```diff
@@ -34,15 +41,36 @@ def __init__(self, logger=None):
         self.logger = logger or setup_logger(self.__class__.__name__)
         self.runner = LocalCommandRunner(logger=self.logger)
 
-    def install(self, source, args=''):
-
+    def install(self, plugin, blueprint_id=None):
         """
         Install the plugin to the current virtualenv.
 
-        :param source: URL to the plugin. Any pip acceptable URL is ok.
-        :param args: extra installation arguments passed to the pip command
+        :param plugin: A plugin structure as defined in the blueprint
         """
+        plugin_id = get_plugin_id(plugin)
+        source = get_plugin_source(plugin, blueprint_id)
+        args = get_plugin_args(plugin)
+        if plugin_id:
+            self._wagon_install(plugin_id, args)
+            return plugin['package_name']
+        elif source:
+            return self._pip_install(source, args)
```

Comment:
as far as I understand, our assumption is that when users declare a plugin in the blueprint, we will look for that plugin and fallback to installation from 'source' only if a source is declared in the blueprint as well. Now let's say that a user declared a plugin and did not intend to install from source.. but they misspelled the version and thus we couldn't find the wagon. In such a care, the plugin_id will be None and we will fallback to source. I believe that in this case, we should at least log a warning stating that the required plugin could not be found and that we're falling back.